### PR TITLE
Y24-065 Create a script to backfill the labware table's new retention_instruction column

### PR DESCRIPTION
**Pull request by dasunpubudumal**
Closes #

#### Changes proposed in this pull request

- 

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: spec/lib/retention_instructions_spec.rb [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                                       create(:custom_metadatum_collection,
+                                              metadata: { 'retention_instruction' => 'Destroy after 2 years' }
+                                       )
+      )
+
+      # Execute
+      run_rake_task
+
+      # Verify
+      expect(labware.reload.retention_instruction).not_to be_nil
```
Comment (BenTopping) [inline spec/lib/retention_instructions_spec.rb:24]: Do you want to check the value of the instruction is correct?

File: spec/lib/retention_instructions_spec.rb [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                                       create(:custom_metadatum_collection,
+                                              metadata: { 'retention_instruction' => 'Destroy after 2 years' }
+                                       )
+      )
+
+      # Execute
+      run_rake_task
+
+      # Verify
+      expect(labware.reload.custom_metadatum_collection.metadata['retention_instruction']).to be_nil
```
Comment (BenTopping) [inline spec/lib/retention_instructions_spec.rb:40]: Can this be included as part of the test above since they are running the same thing?

File: spec/lib/retention_instructions_spec.rb [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                                                'other_key' => 'other_value'
+                                              }
+                                       )
+      )
+
+      # Execute
+      run_rake_task
+
+      # Verify
+      expect(labware.reload.custom_metadatum_collection.metadata['other_key']).to eq('other_value')
```
Comment (BenTopping) [inline spec/lib/retention_instructions_spec.rb:78]: This one and the above are very similar, is the above one needed if this one checks the key exists, which if true means it cant be nil.

File: lib/tasks/retention_instructions.rake [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    labware.retention_instruction = find_retention_instruction_key_for_value(
+      labware.custom_metadatum_collection.metadata['retention_instruction']
+    ).to_sym
+
+    labware.custom_metadatum_collection.custom_metadata.each do |custom_metadata_record|
+      custom_metadata_record.key == 'retention_instruction' && custom_metadata_record.destroy!
+    end
+
+    labware.save! ? saved_count + 1 : saved_count
```
Comment (BenTopping) [inline lib/tasks/retention_instructions.rake:41]: Might need to wrap `custom_metadata_record.destroy!` and `labware.save!` in a begin/rescue as I think it will error on failure instead of returning false.

**Review by BenTopping — COMMENTED** [review]
Haven't tested but looks sensible.

File: spec/lib/retention_instructions_spec.rb (hunk 2 above)
Comment (dasunpubudumal) [reply spec/lib/retention_instructions_spec.rb:40]: Hi Ben, so the test above checks for the `labware.retention_instruction` value, while this test checks whether the scripts removes the retention instruction that is currently stored in `custom_metadata`. So, I think the two tests assert for two different things?

Comment (dasunpubudumal) [reply spec/lib/retention_instructions_spec.rb:40]: Hi Ben, so the test above checks for the `labware.retention_instruction` value, while this test checks whether the scripts removes the retention instruction that is currently stored in `custom_metadata`. So, I think the two tests assert for two different things?

File: spec/lib/retention_instructions_spec.rb (hunk 3 above)
Comment (dasunpubudumal) [reply spec/lib/retention_instructions_spec.rb:78]: So this test intends to check whether the script **does not** remove other metadata (i.e., it checks whether the script removes only the value with the key `retention_instruction` in `custom_metadata` table).

File: lib/tasks/retention_instructions.rake (hunk 4 above)
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:41]: Good spot! Thanks Ben.

File: spec/lib/retention_instructions_spec.rb (hunk 2 above)
Comment (BenTopping) [reply spec/lib/retention_instructions_spec.rb:40]: Yeah thats seems right, even though they assert different things they are outcomes of the same action so could be tested in the same test. That might be personal preference so happy to go with whatever.

File: spec/lib/retention_instructions_spec.rb (hunk 3 above)
Comment (BenTopping) [reply spec/lib/retention_instructions_spec.rb:60]: Again similar to the above, probably different test preferences. If you are testing it does not remove other metadata is that not also testing that metadata is not nil?

File: lib/tasks/retention_instructions.rake (hunk 4 above)
Comment (BenTopping) [reply lib/tasks/retention_instructions.rake]: 👌  Might be worth testing as well but im not sure how easy it is to force validation failures.

File: spec/lib/retention_instructions_spec.rb (hunk 3 above)
Comment (dasunpubudumal) [reply spec/lib/retention_instructions_spec.rb:60]: Yeah I guess you're correct. We could move this assertion into the test above, if that's what you're suggesting? 

I guess I've been following this "one-assertion per test" philosophy, which is [quite old](https://stackoverflow.blog/2022/11/03/multiple-assertions-per-test-are-fine/#h2-459b27e85d780). I need to get out of that one :D

File: spec/lib/retention_instructions_spec.rb [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+      # Execute
+      expect { run_rake_task }.not_to raise_error
+
+      # Verify
+      expect(labware.retention_instruction).to be_nil
+    end
+  end
+
+  context 'when a large number or records are given',
```
Comment (BenTopping) [inline spec/lib/retention_instructions_spec.rb:122]: If we are skipping this can we remove it? Adding a note in the PR to show you have tested it manually but dont want to add it to the test suite is probably fine.

File: spec/lib/retention_instructions_spec.rb (hunk 3 above)
Comment (BenTopping) [reply spec/lib/retention_instructions_spec.rb:50]: Yeah but I see your approach now and both are fine, happy to let you decide.

File: lib/tasks/retention_instructions.rake [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    labware.retention_instruction = find_retention_instruction_key_for_value(
+      labware.custom_metadatum_collection.metadata['retention_instruction']
+    ).to_sym
+
+    begin
+      labware.custom_metadatum_collection.custom_metadata.each do |custom_metadata_record|
+        custom_metadata_record.key == 'retention_instruction' && custom_metadata_record.destroy!
+      end
+      labware.save! ? saved_count + 1 : saved_count
+    rescue ActiveRecord::ActiveRecordError => e
```
Comment (yoldas) [inline lib/tasks/retention_instructions.rake:44]: This exception handling will prevent transaction rollback because the exception will not bubble up to the transaction block. It may be unlikely but If custom_metadata_record.destroy! succeeds and save labware.save! fails, it is a data loss.

Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake]: Agreed. I updated the to fit this.

File: app/sequencescape_excel/sequencescape_excel/specialised_field/retention_instruction.rb
Comment (KatyTaylor) [inline app/sequencescape_excel/sequencescape_excel/specialised_field/retention_instruction.rb:1]: Just wondering... if it's just a normal field on labware now, does it even need to be a 'SpecialisedField'? Or could it just work like a normal field?

File: spec/factories/custom_metadatum_collection_with_asset_factories.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+FactoryBot.define do
+  factory :custom_metadatum_collection_with_asset,  parent: :custom_metadatum_collection do
```
Comment (KatyTaylor) [inline spec/factories/custom_metadatum_collection_with_asset_factories.rb:3]: Sorry... stuck record... can we change `asset` terminology to `labware` please? :)

**Review by KatyTaylor — COMMENTED** [review]
I've added a couple of comments. What's the plan? 

- Deploy and then manually run the rake task on the server?
- Test in training env first - check fixes all the data
- Delete this code afterwards?

File: lib/tasks/retention_instructions.rake [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      labware.custom_metadatum_collection.metadata['retention_instruction'].present?
+
+    labware.retention_instruction = find_retention_instruction_key_for_value(
+      labware.custom_metadatum_collection.metadata['retention_instruction']
+    ).to_sym
+
+    begin
+      labware.custom_metadatum_collection.custom_metadata.each do |custom_metadata_record|
+        custom_metadata_record.key == 'retention_instruction' && custom_metadata_record.destroy!
+      end
```
[the remark this answers is not in the document]
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:41]: Well well well. Not only it is equivalent, but also more efficient 😄 The way I've done it fetches all records (and filters in memory) while the `where` "function" in your suggestion already filters stuff at the database level.

Still getting used to Rails, unfortunately 🚃 😆

[the remark this answers is not in the document]
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:41]: The only difference is that it should be `find_each` instead of each (typical Rubocop stuff).

File: lib/tasks/retention_instructions.rake [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      labware.save! ? saved_count + 1 : saved_count
```
[the remark this answers is not in the document]
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:42]: Agreed. Will update this.

File: spec/factories/custom_metadatum_collection_with_asset_factories.rb (hunk 7 above)
Comment (dasunpubudumal) [reply spec/factories/custom_metadatum_collection_with_asset_factories.rb:3]: This is actually a redundant factory. I'll remove this entire factory.

File: app/sequencescape_excel/sequencescape_excel/specialised_field/retention_instruction.rb
Comment (dasunpubudumal) [reply app/sequencescape_excel/sequencescape_excel/specialised_field/retention_instruction.rb:1]: Interesting. This was already declared as a specialised field. I'm not sure why it was deemed 'specialized' though.

**Comment by dasunpubudumal** [comment]
> I've added a couple of comments. What's the plan?
> 
> * Deploy and then manually run the rake task on the server?
> * Test in training env first - check fixes all the data
> * Delete this code afterwards?

The plan is to actually get this into UAT and to Training, run the task there manually (through SSH) and check how it's going. We'll have to log in to the database and verify whether 1) `retention_instruction` column is populated with the correct data, and whether `custom_metadata` were properly removed. 

In my opinion, we need to export some sort of a restorable database dump in prior to all of these.

File: lib/tasks/retention_instructions.rake [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      labware.custom_metadatum_collection.metadata['retention_instruction'].present?
+
+    labware.retention_instruction = find_retention_instruction_key_for_value(
+      labware.custom_metadatum_collection.metadata['retention_instruction']
+    ).to_sym
+
+    begin
+      labware.custom_metadatum_collection.custom_metadata.where(key: 'retention_instruction').find_each(&:destroy!)
+      saved_count += 1 if labware.save!
+      saved_count
```
Comment (yoldas) [inline lib/tasks/retention_instructions.rake:41]: On failure, labware.save! will raise an exception. On success, it will return true. 

On failure, the last value generated in the block  by the augmented assignment will be nil, but the code jumps to the rescue block. On success, the last value generated in the block will be the result of the augmented assignment.

saved_count statement in the separate line looks redundant.

File: lib/tasks/retention_instructions.rake [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    rescue ActiveRecord::ActiveRecordError => e
+      puts "Error processing labware #{labware.id}: #{e.message}"
+      raise e
+    rescue StandardError => e
+      puts "Error processing labware #{labware.id}: #{e.message}"
+    end
```
Comment (yoldas) [inline lib/tasks/retention_instructions.rake:47]: When can a StandardError happen in this block ? I am asking because we are not rolling back the transaction in this case.

**Review by yoldas — CHANGES_REQUESTED** [review]

File: lib/tasks/retention_instructions.rake (hunk 11 above)
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:47]: Yeah I think it should also raise. I added the standard error to make it more generalisable. I didn't actually find a case where it explicitly throw a `StandardError` though.

File: lib/tasks/retention_instructions.rake (hunk 10 above)
Comment (dasunpubudumal) [reply lib/tasks/retention_instructions.rake:41]: So the idea is to return the updated count after the script is done without errors. So I think it needs to be returned from this function, right? Because only incrementing would not suffice as the saved_count initialized within the transaction is copied into this function's stack. We can return it at the end of the function though.

**Review by yoldas — APPROVED** [review]

File: lib/tasks/retention_instructions.rake [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    args.with_defaults(batch_size: 1000, limit: nil)
+    batch_size = args[:batch_size].to_i
+    limit = args[:limit]&.to_i
+
+    puts "Backfilling retention instructions with batch size: #{batch_size}..."
+
+    ActiveRecord::Base.transaction do
+      saved_count = 0
+      labwares = Labware.where(retention_instruction: nil)
+      labwares = labwares.limit(limit) unless limit.nil?
```
Comment (dasunpubudumal) [inline lib/tasks/retention_instructions.rake:21]: @yoldas A `limit` parameter just to test the script in UAT. This is by default `nil` which prompts to fetch all the records.

Comment (yoldas) [reply lib/tasks/retention_instructions.rake:21]: You could add `.order(:id)` to get the labwares in a predictable order to help testing and to quickly know where you would start next time. It is a good idea and it looks good otherwise.

**Review by yoldas — APPROVED** [review]

File: lib/tasks/retention_instructions.rake (hunk 11 above)
Comment (yoldas) [reply lib/tasks/retention_instructions.rake:48]: I see you have now re-throwing the exception in that block similar to the other. I am resolving this convo.

File: lib/tasks/retention_instructions.rake [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  # rubocop:todo Metrics/AbcSize
+  # rubocop:todo Metrics/MethodLength
+  def process_labware(labware, saved_count)
+    retention_instruction = find_retention_instruction_key_for_value(
+      labware.custom_metadatum_collection.metadata['retention_instruction']
+    )
+
+    # Had to add this because there are some labwares with retention instructions like 'not provided'!
```
Comment (dasunpubudumal) [inline lib/tasks/retention_instructions.rake:39]: `not provided` thing @yoldas

**Review by yoldas — APPROVED** [review]
